### PR TITLE
Autodiscover

### PR DESCRIPTION
**Pull request by exekias**
This PR adds support for Autodiscover, and implements a first provider for Docker.

### Autodiscover

Autodiscover allows the user to define different providers to watch for system changes and instantiate new modules as these happen.

Providers watch for changes and emit events to a common bus, then the autodiscovery process detect situations when there is something we can monitor and instantiate a modules for it.

Bus messages contain useful information about what providers see, with some common well-known optional fields (provider specific fields may also be present):

- host - service host
- port - service port
- config - recommended config for this instance (from the user or environment)

#### Docker provider

Docker provider watches for docker container events, it supports config mapping from container metadata -> config templates, so new modules are created when a container starts. The available data for a container (that you can use in the config template) is:

- host
- port
- docker.container.id
- docker.container.name
- docker.container.image
- docker.container.labels

#### Testing

This should be enough to monitor any present and future redis instances that run on docker:
```
metricbeat.autodiscover:
  providers:
    - type: docker
      templates:
        - condition:
            equals.docker.container.image: redis
          config:
            - module: redis
              metricsets: ["info", "keyspace"]
              hosts: "${data.host}:${data.port}"
```

Or reading redis logs only:
```
filebeat.autodiscover:
  providers:
    - type: docker
      templates:
        - condition:
            equals.docker.container.image: redis
          config:
            - type: log
              paths:
                - /var/lib/docker/containers/${data.docker.container.id}/*.log
```

You can launch it like this:
```
docker run -p 6379:6379 redis
```
This PR is in progress, with some missing parts:
- [x] Revisit Docker IP/port mapping, do we really need exposed ports? probably not
- [x] Add support in filebeat
- [x] Revisit bus implementation, use interfaces (instead of MapStr)?
- [x] Add docker metadata to modules discovered by docker provider
- [x] System tests
- [x] Update metricbeat and filebeat YAML files
- [ ] Docs

### CODE REVIEW

File: libbeat/autodiscover/provider.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+// NewRegistry creates and returns a new Registry
+func NewRegistry() *registry {
+	return &registry{
+		providers: make(map[string]ProviderBuilder, 0),
+	}
+}
+
+// AddProvider registers a new ProviderBuilder
+func (r *registry) AddProvider(name string, provider ProviderBuilder) error {
+	r.lock.Lock()
```
Comment (kvch) [inline libbeat/autodiscover/provider.go]: Why do you lock here? Are `init` functions called concurrently?

File: libbeat/common/bus/bus.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// Event sent to the bus
+type Event common.MapStr
+
+// Bus provides a common channel to emit and listen for Events
+type Bus interface {
+	// Emit an event to the bus
+	Emit(Event)
```
Comment (kvch) [inline libbeat/common/bus/bus.go]: I think `Publish` would be a more expressive name than `Emit`, as it is a kind of Publish-Subscribe model.

File: libbeat/autodiscover/common/config.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
+package common
```
Comment (ruflin) [inline libbeat/autodiscover/common/config.go]: I worry a bit that this will conflict with our libbeat common package in some places.

File: libbeat/autodiscover/providers/docker/config.go [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package docker
+
+import (
+	"github.com/elastic/beats/libbeat/autodiscover/common"
+	"github.com/elastic/beats/libbeat/processors/add_docker_metadata"
```
Comment (ruflin) [inline libbeat/autodiscover/providers/docker/config.go]: Feels strange to include here a processor. Perhaps we should extract the TLS config part to a docker package.

File: libbeat/autodiscover/providers/docker/docker.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
+package docker
+
+import (
+	"github.com/elastic/beats/libbeat/autodiscover"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/bus"
+	"github.com/elastic/beats/libbeat/logp"
+
+	// TODO move add_docker_metadata.Watcher to a common place
```
Comment (ruflin) [inline libbeat/autodiscover/providers/docker/docker.go]: :-D

File: libbeat/processors/add_docker_metadata/watcher.go [hunk 6]
```diff
@@ -32,6 +33,12 @@ type Watcher interface {
 
 	// Containers returns the list of known containers
 	Containers() map[string]*Container
+
+	// ListenStart returns a bus listener to receive container started events, with a `container` key holding it
+	ListenStart() bus.Listener
```
Comment (ruflin) [inline libbeat/processors/add_docker_metadata/watcher.go]: I wonder if we should put these in the same interface or one more specific to autodiscovery which is composeo f this Watcher interface.

**Comment by ruflin** [comment]
@exekias I stumbled over the config a bit because of all the indentation it has. Here an alternative suggestion:

```
metricbeat.autodiscover:
  providers:
    - type: docker
      filter:
        docker-image: redis
        port: 6309
      config:
        - module: redis
          metricsets: ["info"]
          hosts: "${data.host}:${data.port}"
        - module: redis
          metricsets: ["info", "keyspace"]
          hosts: "${data.host}:${data.port}"
          period: 60s
    - type: docker
      conditions:
        - when.equals.docker-image: mysql
        - when.range.port.gte: 6000
      config:
        - module: mysql
          metricsets: ["stats"]
          hosts: "${data.host}:${data.port}"
```

The basic idea is to make it more obvious out of what the config will be generated and allow multiple `and` criterias / filters in a simple way. But at the same time provide an option to use our more powerful conditions coming from the processors for matching. 

I don't think we should provide the conditions in the first implementation but wanted to add it here to take into account where this could go. I'm not happy with `filter` naming in the first config part. Perhaps `criteria` or `match` could be an alternative?

I'm pretty sure the above config brings up other issues as I only focused on the config and not potential implications on the implementation itself.

File: libbeat/autodiscover/common/config.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+					logp.Err("Error building config: %v", err)
+				}
+				opts := []ucfg.Option{
+					ucfg.PathSep("."),
+					ucfg.Env(vars),
+					ucfg.ResolveEnv,
+					ucfg.VarExp,
+				}
+				for _, config := range configs {
+					c, err := ucfg.NewFrom(config, opts...)
```
Comment (urso) [inline libbeat/autodiscover/common/config.go]: Why do you need ucfg here?

Minor problem with this approach is, you are loosing the full setting paths. If some config unpacking fails, the setting it's full path is reported with the error message.

File: libbeat/autodiscover/provider.go (hunk 1 above)
Comment (exekias) [reply libbeat/autodiscover/provider.go]: Nope, or not at the moment :innocent:, but some defensive code there won't hurt I think, this may be a thing in the future?

File: libbeat/autodiscover/common/config.go (hunk 7 above)
Comment (exekias) [reply libbeat/autodiscover/common/config.go]: I pass provider event data for variable expansion, so config templates can use data from the event. Agree to the minor issue (would affect logs mainly?)

**Comment by vjsamuel** [comment]
@exekias on general thought. with autodiscover we should avoid using `add_docker_metadata` and `add_kubernetes_metadata`, we should provide all metadata upfront as part of the module configuration. This will improve performance tremendously as we would not need to do append of metadata at each event.

we should also have some plugability in place for folks who want to run something custom on top of auto discovery. this is similar to collectbeat appenders where we have our own internal appenders more custom tailored for our needs.

File: libbeat/autodiscover/common/config.go (hunk 7 above)
Comment (urso) [reply libbeat/autodiscover/common/config.go]: Yes, it's for logs only. As we can't store file positions (by line) yet, not having the correct names can make it more difficult to track-down misconfigurations.

Comment (vjsamuel) [reply libbeat/autodiscover/common/config.go]: discovery can be more automatic than providing `docker-image: redis`. if we use docker labels, we should be able to pass `annotation` like information same like collectbeat to be able to discover workloads automatically. i could run 10 containers in a host. it would be a huge config if i have to say for each via config how it has to be discovered.

Comment (exekias) [reply libbeat/autodiscover/common/config.go]: I agree @vjsamuel, this is a first PR, intended to bring the basic functionality and building blocks to grow from there. Definitely that's something we want to have

**Comment by fin09pcap** [comment]
subscribe

**Comment by exekias** [comment]
@ruflin I updated config format to take advantage of processor conditions, let me know what you think about it:
```
metricbeat.autodiscover:
  providers:
    - type: docker
      templates:
        - when.equals.docker-image: redis
          use:
            - module: redis
              metricsets: ["info", "keyspace"]
              hosts: "${data.host}:${data.port}"
```

**Comment by christiangalsterer** [comment]
This is a really great feature as this really great for mid to large scale Docker/Kubernetes installations to auto-configure Beats when container come and go.
Do you have by chance any idea in which version and when this will be available? A first indication would already help if this is something we can expect in the next month, next quarter, etc.

**Comment by exekias** [comment]
Hi @christiangalsterer, I'm glad you like it! this is something we want to have in the 6.X timeline, probably in a few months

**Comment by ruflin** [comment]
We had a discussion with @exekias and @urso about the config and that is what we came up with. I added some comments to explain why the decision was made this way:

```
metricbeat.autodiscover:
  # Each Provider can be defined multiple times.
  # Example: Assuming we have a port scanner provider, different scan intervals are applied for different port ranges
  providers:
    - type: docker
    
      # Each provider can have it's specific config options
      host: docker/socket/path
      
      # Multiple templates can be used per provider with different condtions
      templates:
      
          # One condition is used per template. Sub conditions can be added with and to nest conditions
        - condition:
            and:
              - equals.docker.container.image: redis
        
          # The configuration fields that will be created
          config:
            - module: redis
              metricsets: ["info"]
              hosts: "${data.host}:${data.port}"
              frequency: 1m
            - module: redis
              metricsets: ["keyspace"]
              hosts: "${data.host}:${data.port}"
              frequency: 10s
              
          # Same config as above just starting with config instead of condition
        - config:
            - module: redis
              metricsets: ["info", "keyspace"]
              hosts: "${data.host}:${data.port}"
          condition:
            when.equals.docker.container.image: redis
```

We couldn't decide if it should be called `config` or `configs` but sticked with `config` for now.

File: filebeat/crawler/crawler.go [hunk 8]
```diff
@@ -21,6 +21,8 @@ type Crawler struct {
 	prospectorConfigs   []*common.Config
 	out                 channel.Factory
 	wg                  sync.WaitGroup
+	ProspectorsFactory  cfgfile.RunnerFactory
```
Comment (ruflin) [inline filebeat/crawler/crawler.go]: Note for myself: Need to understand why we need to store the factory now in the Crawler struct.

File: filebeat/fileset/factory.go [hunk 9]
```diff
@@ -110,6 +110,7 @@ func (p *prospectorsRunner) Stop() {
 		prospector.Stop()
 	}
 }
-func (p *prospectorsRunner) ID() uint64 {
```
Comment (ruflin) [inline filebeat/fileset/factory.go]: Was the ID still used somewhere? If not, perhaps remove it in a different PR?

File: filebeat/prospector/prospector.go [hunk 10]
```diff
@@ -141,3 +144,7 @@ func (p *Prospector) stop() {
 		p.prospectorer.Stop()
 	}
 }
+
+func (p *Prospector) String() string {
```
Comment (ruflin) [inline filebeat/prospector/prospector.go]: Is that a required method or just for convenience / logging?

File: libbeat/autodiscover/autodiscover.go [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+	// CreateConfig generates a valid list of configs from the given event, the received event will have all keys defined by `StartFilter`
+	CreateConfig(bus.Event) ([]*common.Config, error)
+
+	// CheckConfig tests given config to check if it will work or not, returns errors in case it won't work
+	CheckConfig(*common.Config) error
+
+	// RunnerFactory provides runner creation by feeding valid configs
+	cfgfile.RunnerFactory
+
+	// StartFilter returns the bus filter to retrieve runner start triggering events
```
Comment (ruflin) [inline libbeat/autodiscover/autodiscover.go]: So far it seems these are all the same for the current implementation. How will these look for other adapters?

File: libbeat/autodiscover/autodiscover.go [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+		bus:       bus,
+		adapter:   adapter,
+		runners:   cfgfile.NewRegistry(),
+		providers: providers,
+		meta:      make(map[uint64]*common.MapStrPointer),
+	}, nil
+}
+
+// Start autodiscovery process
+func (a *Autodiscover) Start() {
```
Comment (ruflin) [inline libbeat/autodiscover/autodiscover.go]: Some idea that popped up in my head when reading this: Will it be needed to enabled / disable autodiscovery dynamically? For example on startup you want the beat to autodiscover all the services that are out there and then stop the discovery for example? Or only use autodiscovery for some time enabled through centralised config? We definitively not need that now, just thinking out loud.

File: libbeat/autodiscover/autodiscover.go [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+				runner.Stop()
+				a.runners.Remove(hash)
+			} else {
+				logp.Debug(debugK, "Runner not found for stopping: %s", hash)
+			}
+		}
+	}
+}
+
+func getMeta(event bus.Event) *common.MapStr {
```
Comment (ruflin) [inline libbeat/autodiscover/autodiscover.go]: Should this be a method on `bus.Event`?

File: libbeat/autodiscover/autodiscover.go [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+
+func (a *Autodiscover) storeMeta(hash uint64, meta *common.MapStr) *common.MapStrPointer {
+	a.metaMutex.Lock()
+	defer a.metaMutex.Unlock()
+
+	if meta == nil {
+		return nil
+	}
+
+	if a.meta[hash] == nil {
```
Comment (ruflin) [inline libbeat/autodiscover/autodiscover.go]: I wonder if the code would become simpler if `MetaList` or similar would be it's own struct so all these methods and locks would be on metaList instead of the Autodiscovery object.

File: libbeat/autodiscover/provider.go [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// Provider for autodiscovery
+type Provider interface {
+	cfgfile.Runner
+}
+
+// Registry holds all known autodiscover providers, they must be added to it to enable them for use
+var Registry = NewRegistry()
```
Comment (ruflin) [inline libbeat/autodiscover/provider.go]: Should we call it ProviderRegistry?

File: libbeat/autodiscover/provider.go [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+	logp.Debug(debugK, "Provider registered: %s", name)
+	return nil
+}
+
+// GetProvider returns the provider with the giving name, nil if it doesn't exist
+func (r *registry) GetProvider(name string) ProviderBuilder {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+
+	name = strings.ToLower(name)
```
Comment (ruflin) [inline libbeat/autodiscover/provider.go]: What is the reason you do the `ToLower`?

File: libbeat/autodiscover/providers/docker/config.go [hunk 17]
```diff
@@ -0,0 +1,18 @@
+package docker
```
Comment (ruflin) [inline libbeat/autodiscover/providers/docker/config.go]: I would call the package path `autodiscovery/provider/docker/*` instead of `providers`. It's messy across the code base but I think recently we tried to call it plural if it has multiple implementations directly in the package itself and singular if it has multiple sub packages with one implementation each.

File: libbeat/autodiscover/providers/docker/docker.go [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+			"name":   container.Name,
+			"image":  container.Image,
+			"labels": container.Labels,
+		},
+	}
+
+	// Emit container info
+	d.emit(bus.Event{
+		flag:     true,
+		"host":   container.IPAddresses[0],
```
Comment (ruflin) [inline libbeat/autodiscover/providers/docker/docker.go]: can it happen that there is no IPAdress -> panic? Just double checking.

File: libbeat/autodiscover/providers/docker/docker.go [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+		"meta": common.MapStr{
+			"docker": meta,
+		},
+	})
+
+	// Emit container private ports
+	for _, port := range container.Ports {
+		event := bus.Event{
+			flag:     true,
+			"host":   container.IPAddresses[0],
```
Comment (ruflin) [inline libbeat/autodiscover/providers/docker/docker.go]: see above

File: libbeat/cfgfile/registry.go [hunk 20]
```diff
@@ -3,7 +3,7 @@ package cfgfile
 import "sync"
 
 type Registry struct {
-	sync.Mutex
+	sync.RWMutex
```
Comment (ruflin) [inline libbeat/cfgfile/registry.go]: I like that change. Could we move it to a separate PR to get it in quickly? Trying to reduce the size of this PR :-)

File: libbeat/common/bus/bus.go [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+
+	// Subscribe to all events, filter them to the ones containing *all* the keys in filter
+	Subscribe(filter ...string) Listener
+}
+
+// Listener retrieves Events from a Bus subscription until Close is called
```
Comment (ruflin) [inline libbeat/common/bus/bus.go]: until Stop is called?

File: libbeat/common/bus/bus.go [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+type listener struct {
+	filter  []string
+	channel chan Event
+	bus     *bus
+}
+
+// New initializes a new bus with the given name and returns it
+func New(name string) Bus {
```
Comment (ruflin) [inline libbeat/common/bus/bus.go]: What are the names of the buses you have in mind?

File: libbeat/common/docker/watcher_test.go [hunk 23]
```diff
@@ -12,15 +12,17 @@ import (
...
+	containers [][]types.Container
 	// event list to send on Events call
 	events []interface{}
 
 	done chan interface{}
 }
 
 func (m *MockClient) ContainerList(ctx context.Context, options types.ContainerListOptions) ([]types.Container, error) {
-	return m.containers, nil
+	res := m.containers[0]
```
Comment (ruflin) [inline libbeat/common/docker/watcher_test.go]: Can you elaborate on this change?

File: libbeat/common/mapstr_pointer.go [hunk 24]
```diff
@@ -0,0 +1,29 @@
+package common
+
+import (
+	"sync/atomic"
+	"unsafe"
+)
+
+// MapStrPointer stores a pointer to atomically get/set a MapStr object
```
Comment (ruflin) [inline libbeat/common/mapstr_pointer.go]: I would have prefered to have "optimizations" like this in a follow up PR to have the code we put in as simple as possible to start with and then when we do optimizations we also add some benchmark tests for it. Makes it easier to later detect in case these optimizations causes some issues, if it happened with the optimization or the original code.

File: libbeat/processors/condition.go [hunk 25]
```diff
@@ -42,6 +42,10 @@ type WhenProcessor struct {
 	p         Processor
 }
 
+type ValuesMap interface {
```
Comment (ruflin) [inline libbeat/processors/condition.go]: Note to myself: Figure out why this change is needed and if it could happen in a separate PR.

File: filebeat/crawler/crawler.go (hunk 8 above)
Comment (exekias) [reply filebeat/crawler/crawler.go]: Short answer: we use this factory from Autodiscovery. In the long term, we should unify all prospector/module initialization in a common place

File: filebeat/fileset/factory.go (hunk 9 above)
Comment (exekias) [reply filebeat/fileset/factory.go]: It wasn't used, it's a spurious implementation, probably forgot from a previous removal

File: filebeat/prospector/prospector.go (hunk 10 above)
Comment (exekias) [reply filebeat/prospector/prospector.go]: It's for logging purposes only, autodiscover will tell about new spawned prospectors/modules in the log output

File: libbeat/autodiscover/autodiscover.go (hunk 11 above)
Comment (exekias) [reply libbeat/autodiscover/autodiscover.go]: I don't know yet :innocent:, I think this code can be used in other parts of beats, not only to detect new modules to lauch, so wanted to keep this open

File: libbeat/autodiscover/autodiscover.go (hunk 12 above)
Comment (exekias) [reply libbeat/autodiscover/autodiscover.go]: this is a good point, something like modules reload could make sense for autodiscover providers

File: libbeat/autodiscover/autodiscover.go (hunk 13 above)
Comment (exekias) [reply libbeat/autodiscover/autodiscover.go]: bus.Event is quite generic and can be used in other parts, `meta` field is just a convention for autodiscover. I would prefer to keep bus slim until we see how it gets used

File: libbeat/autodiscover/autodiscover.go (hunk 14 above)
Comment (exekias) [reply libbeat/autodiscover/autodiscover.go]: Totally, will do it, thanks!

File: libbeat/autodiscover/provider.go (hunk 16 above)
Comment (exekias) [reply libbeat/autodiscover/provider.go]: User input sanitizing, `type: Docker` would still work, I think I copied this from somewhere else but I'm not sure :P

File: libbeat/autodiscover/providers/docker/docker.go (hunk 18 above)
Comment (exekias) [reply libbeat/autodiscover/providers/docker/docker.go]: good point, It should not happen, but will add some defensive code here

File: libbeat/common/bus/bus.go (hunk 22 above)
Comment (exekias) [reply libbeat/common/bus/bus.go]: I don't know, this name is used just for logging

File: libbeat/common/docker/watcher_test.go (hunk 23 above)
Comment (exekias) [reply libbeat/common/docker/watcher_test.go]: I added support for several requests to this mock, each request returns the next response and pops it out the list

File: libbeat/common/mapstr_pointer.go (hunk 24 above)
Comment (exekias) [reply libbeat/common/mapstr_pointer.go]: Agreed, but as I try to get a stable API to introduce new providers as soon as possible this somehow made sense while I waited for reviews. Once this is merged, autodiscover API should have everything we need

File: libbeat/autodiscover/autodiscover.go (hunk 11 above)
Comment (urso) [reply libbeat/autodiscover/autodiscover.go]: Do we even need the filters methods on the interface? Can the filters be passed when registering the adapter, so to keep the interface somewhat 'smaller'?

File: libbeat/beat/pipeline.go [hunk 26]
```diff
@@ -34,6 +34,9 @@ type ClientConfig struct {
 	// Fields provides additional 'global' fields to be added to every event
 	Fields common.MapStr
 
+	// DynamicMeta provides additional fields to be added to every event, supporting live updates
+	DynamicMeta *common.MapStrPointer
```
Comment (urso) [inline libbeat/beat/pipeline.go]: To reduce confusion, rather use `DynamicFields`. The beat event has `Meta` and `Fields`. The `Meta` are somewhat internal (like LS `@metadata`). But `DynamicMeta` is updating the event `Fields`, not `Meta`.

Btw. instead of adding the functionality to the publisher pipeline, you can just pass a processor doing the magic.

Instead of passing DynamicMeta, we could add atomic values support to the normalizer of `common.MapStr`. This would allow you to pass updatable fields via `Fields` (given you don't need them on top-level).

File: libbeat/common/mapstr_pointer.go [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+// This should give faster access for use cases with lots of reads and a few
+// changes.
+// It's imortant to note that modifying the map is not thread safe, only fully
+// replacing it.
+type MapStrPointer struct {
+	p unsafe.Pointer
+}
```
Comment (urso) [inline libbeat/common/mapstr_pointer.go]: How about changing the type to:

```
type VolatileMapStr struct {
  p *unsafe.Pointer
}

func MakeVolatileMapStr(m MapStr) VolatileMapStr {
  return VolatileMapStr{ p: &unsafe.Pointer(&m) }
}

func (m VolatileMapStr) Get() MapStr {
  return *((*MapStr)(atomic.LoadPointer(m.p)))
}

func (m VolatileMapStr) Set(new MapStr) {
  atomic.StorePointer(m.p, unsafe.Pointer(&new))
}

```

Then you can use `VolatileMapStr` and `MapStr` by value, just like `MapStr` is already passed by value.

File: libbeat/beat/pipeline.go (hunk 26 above)
Comment (exekias) [reply libbeat/beat/pipeline.go]: Thanks for the input, I agree DynamicFields is the right name.

I thought about the processor but that would cause Copy to be called for all cases, I guess that's not too bad?

File: libbeat/common/mapstr_pointer.go (hunk 27 above)
Comment (exekias) [reply libbeat/common/mapstr_pointer.go]: LGTM, but don't really like the Volatile name, somehow it hides the real thing?

File: libbeat/autodiscover/providers/docker/config.go (hunk 17 above)
Comment (exekias) [reply libbeat/autodiscover/providers/docker/config.go]: There is something I don't like from that approach. If I'm the top folder of these layout and I find `autodiscover/provider` I don't realize I have several different providers inside. What's the benefit of using singular for this case?

Comment (ruflin) [reply libbeat/autodiscover/providers/docker/config.go]: If you read the full package name it reads that docker is only 1 provider. But this is definitively personal preference and something @urso and I came up with over time. Not something that would hold back this PR :-)

File: libbeat/autodiscover/autodiscover.go (hunk 11 above)
Comment (exekias) [reply libbeat/autodiscover/autodiscover.go]: The adapter is not registered but passed to Autodiscover Manager. If you look at is implementation, it's very small generally, I think it makes sense to keep these filters together with the functions that use it. If in the future we use something different from config we need to update the filters too https://github.com/exekias/beats/blob/411d13a878d64dbe62b6008057bbaa1c207f9e4e/metricbeat/beater/autodiscover.go

File: libbeat/autodiscover/provider.go (hunk 16 above)
Comment (ruflin) [reply libbeat/autodiscover/provider.go]: I would suggest not to sanatise it and force users to do "correct" config files instead. Reducing magic :-)

Comment (exekias) [reply libbeat/autodiscover/provider.go]: Just found where I copied this from: https://github.com/elastic/beats/blob/354fdd6792b77ed5f099e017caab7c8ec4fa2ead/metricbeat/mb/registry.go#L77, I'm ok with removing it anyway

File: libbeat/cfgfile/registry.go (hunk 20 above)
Comment (kvch) [reply libbeat/cfgfile/registry.go]: huge :+1: on reducing the size of this PR.

File: metricbeat/metricbeat.yml [hunk 28]
```diff
@@ -19,6 +19,24 @@ metricbeat.config.modules:
   # Period on which files under path should be checked for changes
   #reload.period: 10s
 
+#============================== Autodiscover ===================================
+
+# Autodiscover allows you to detect changes in the system and spawn new modules
+# as they happen.
+
+metricbeat.autodiscover:
```
Comment (tsg) [inline metricbeat/metricbeat.yml]: I think Autodiscovery doesn't need to be in the short config (which we try to keep minimal), but only in the reference one. Especially since Autodiscovery is going to be in the K8s deployment manifersts, right?

Comment (exekias) [reply metricbeat/metricbeat.yml]: :+1: that works for me

**Comment by exekias** [comment]
jenkins retest it

**Comment by exekias** [comment]
This PR is ready for a (final?) review, I'm currently opening side PRs to shrink this one

1. https://github.com/elastic/beats/pull/5581

**Comment by ruflin** [comment]
@exekias just merged the other two. Could you rebase on top? If it reduces the hassle of rebasing, feel free to squash it into 1 commit already (or as many as you want in the end).

**Comment by tsg** [comment]
@exekias Merged, but please keep in mind to add docs before the 6.1 GA.

File: libbeat/autodiscover/provider.go (hunk 16 above)
Comment (ruflin) [reply libbeat/autodiscover/provider.go]: Undecided on this one. Probably easier on our users. @andrewkroh WDYT?

File: libbeat/autodiscover/providers/docker/config.go (hunk 17 above)
Comment (ruflin) [reply libbeat/autodiscover/providers/docker/config.go]: @exekias you sneaked it in ;-)

**Comment by ruflin** [comment]
@exekias Awesome stuff. Thanks for making this happen.

File: libbeat/autodiscover/provider.go (hunk 16 above)
Comment (andrewkroh) [reply libbeat/autodiscover/provider.go]: As the person who made the registry case-insensitive I like that it accepts any case as input.

**Comment by dedemorton** [comment]
Autodiscover docs were added with https://github.com/elastic/beats/pull/5868, so I'm removing the needs_docs label.
